--- a/net/netfilter/nf_conntrack_proto_tcp.c	2018-01-29 14:00:50.394586377 -0800
+++ b/net/netfilter/nf_conntrack_proto_tcp.c	2017-09-05 10:37:46.000000000 -0700
@@ -527,12 +527,17 @@ static bool tcp_in_window(const struct n
 		return true;
 
 	if (net) {
-		if ((net->ipv4.devconf_all && net->ipv4.devconf_dflt && net->ipv6.devconf_all) &&
-		    net->ipv6.devconf_dflt) {
+		if ((net->ipv4.devconf_all && net->ipv4.devconf_dflt)
+#ifdef IPV6
+ && (net->ipv6.devconf_all && net->ipv6.devconf_dflt)
+#endif
+		) {
 			if ((IPV4_DEVCONF_DFLT(net, FORWARDING) ||
-			     IPV4_DEVCONF_ALL(net, FORWARDING)) ||
-			     (net->ipv6.devconf_all->forwarding ||
-			      net->ipv6.devconf_dflt->forwarding)) {
+			     IPV4_DEVCONF_ALL(net, FORWARDING)) 
+#ifdef IPV6
+|| (net->ipv6.devconf_all->forwarding || net->ipv6.devconf_dflt->forwarding)
+#endif
+			) {
 				return true;
 			}
 		}
