--- a/net/net.c	2018-01-23 15:46:57.380144246 -0800
+++ b/net/net.c	2018-01-23 15:58:17.124136867 -0800
@@ -181,7 +181,7 @@ IPaddr_t	NetNtpServerIP;
 int		NetTimeOffset;
 #endif
 
-static uchar PktBuf[(PKTBUFSRX+1) * PKTSIZE_ALIGN + PKTALIGN];
+volatile uchar PktBuf[(PKTBUFSRX+1) * PKTSIZE_ALIGN + PKTALIGN];
 
 /* Receive packet */
 uchar *NetRxPackets[PKTBUFSRX];
@@ -301,7 +301,7 @@ void net_init(void)
 		 */
 		int i;
 
-		NetTxPacket = &PktBuf[0] + (PKTALIGN - 1);
+		NetTxPacket = (uchar *)(&PktBuf[0] + (PKTALIGN - 1));
 		NetTxPacket -= (ulong)NetTxPacket % PKTALIGN;
 		for (i = 0; i < PKTBUFSRX; i++)
 			NetRxPackets[i] = NetTxPacket + (i + 1) * PKTSIZE_ALIGN;
--- a/include/net.h	2018-01-23 15:46:57.376144246 -0800
+++ b/include/net.h	2018-01-23 15:44:30.540145840 -0800
@@ -85,7 +85,7 @@ struct eth_device {
 	int state;
 
 	int  (*init) (struct eth_device *, bd_t *);
-	int  (*send) (struct eth_device *, void *packet, int length);
+	int  (*send) (struct eth_device *, volatile void *packet, int length);
 	int  (*recv) (struct eth_device *);
 	void (*halt) (struct eth_device *);
 #ifdef CONFIG_MCAST_TFTP
--- a/drivers/net/Makefile	2018-01-23 15:46:57.348144246 -0800
+++ b/drivers/net/Makefile	2018-01-23 15:44:30.896145836 -0800
@@ -63,3 +63,5 @@ obj-$(CONFIG_XILINX_EMACLITE) += xilinx_
 obj-$(CONFIG_XILINX_LL_TEMAC) += xilinx_ll_temac.o xilinx_ll_temac_mdio.o \
 		xilinx_ll_temac_fifo.o xilinx_ll_temac_sdma.o
 obj-$(CONFIG_ZYNQ_GEM) += zynq_gem.o
+obj-$(CONFIG_RT2880_ETH) += rt2880_eth.o
+obj-$(CONFIG_RT2880_ETH) += mii_mgr.o
