--- a/core/notifier.c	2018-04-24 09:54:47.982908811 -0700
+++ b/core/notifier.c	2018-04-24 09:55:36.415430532 -0700
@@ -227,6 +227,7 @@ static void *notifier_thread (void __att
 {
 	int serverfd;
 	int len;
+	int sock_optval = 1;
 	struct notify_ipc_msg msg;

 	/* Initialize and bind to UDS */
@@ -236,6 +237,13 @@ static void *notifier_thread (void __att
 		exit(2);
 	}

+	if ((setsockopt(serverfd, SOL_SOCKET, SO_REUSEADDR,
+					 &sock_optval, sizeof(sock_optval))) < 0) {
+		TRACE("Error setting socket options: %s", strerror(errno));
+		close(serverfd);
+		exit(2);
+	}
+
 	if (bind(serverfd, (const struct sockaddr *) &notify_server,
 			sizeof(struct sockaddr_un)) < 0) {
 		fprintf(stderr, "Error binding notifier socket: %s, exiting.\n", strerror(errno));
